modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1127 @@
+          {
+            dof_id_type node3 = cut_elem2->get_node(k)->id();
+            dof_id_type node4 = cut_elem2->get_node((k + 1 <= 2) ? k + 1 : 0)->id();
+            if (node3 > node4)
+              std::swap(node3, node4);
+
+            if (node1 == node3 && node2 == node4)
+            {
+              is_edge_inside = 1;
+              goto endloop;
+            }
+          }
+        }
+        ++counter;
+      }
+    endloop:
+      if (is_edge_inside == 0)
+      {
+        // store boundary edges
+        CutEdge ce;
+        ce.id1 = node1;
+        ce.id2 = node2;
+        _boundary_edges.insert(ce);
+      }
+      else
+      {
+        // this is not a boundary edge; remove it from existing edge list
+        for (auto it = _boundary_edges.begin(); it != _boundary_edges.end();)
+        {
+          if ((*it).id1 == node1 && (*it).id2 == node2)
+            it = _boundary_edges.erase(it);
+          else
+            ++it;
+        }
+      }
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::sortBoundaryNodes()
+{
+  _boundary.clear();
+
+  for (auto it = _boundary_edges.begin(); it != _boundary_edges.end(); ++it)
+  {
+    dof_id_type node1 = (*it).id1;
+    dof_id_type node2 = (*it).id2;
+
+    mooseAssert(_boundary_map.find(node1) != _boundary_map.end(),
+                "_boundary_map does not have this key");
+    mooseAssert(_boundary_map.find(node2) != _boundary_map.end(),
+                "_boundary_map does not have this key");
+
+    _boundary_map.find(node1)->second.push_back(node2);
+    _boundary_map.find(node2)->second.push_back(node1);
+  }
+
+  auto it = _boundary_map.begin();
+  while (it != _boundary_map.end())
+  {
+    if (it->second.size() != 2)
+      mooseError(
+          "Boundary nodes in the cutter mesh must have exactly two neighbors; this one has: ",
+          it->second.size());
+    ++it;
+  }
+
+  auto it2 = _boundary_edges.begin();
+  dof_id_type node1 = (*it2).id1;
+  dof_id_type node2 = (*it2).id2;
+  _boundary.push_back(node1);
+  _boundary.push_back(node2);
+
+  for (unsigned int i = 0; i < _boundary_edges.size() - 1; ++i)
+  {
+    mooseAssert(_boundary_map.find(node2) != _boundary_map.end(),
+                "_boundary_map does not have this key");
+
+    dof_id_type node3 = _boundary_map.find(node2)->second[0];
+    dof_id_type node4 = _boundary_map.find(node2)->second[1];
+
+    if (node3 == node1)
+    {
+      _boundary.push_back(node4);
+      node1 = node2;
+      node2 = node4;
+    }
+    else if (node4 == node1)
+    {
+      _boundary.push_back(node3);
+      node1 = node2;
+      node2 = node3;
+    }
+    else
+      mooseError("Discontinuity in cutter boundary");
+  }
+}
+
+Real
+MeshCut3DUserObject::findDistance(dof_id_type node1, dof_id_type node2)
+{
+  Node * n1 = _cut_mesh->node_ptr(node1);
+  mooseAssert(n1 != nullptr, "Node is NULL");
+  Point & p1 = *n1;
+  Node * n2 = _cut_mesh->node_ptr(node2);
+  mooseAssert(n2 != nullptr, "Node is NULL");
+  Point & p2 = *n2;
```

Comment:
No need to store into a Point. Nodes can do norms too. Also, the reason you'd get a nullptr is because of DistributedMesh. I like the asserts!

Reply:
Done.